sky/data/storage.py

```diff
@@ -1074,13 +1138,22 @@ def upload(self):
             StorageUploadError: if upload fails.
         """
         try:
-            if self.source is not None:
+            if isinstance(self.source, list):
+                for source_path in self.source:
+                    # We upload each file in the list as individually because
+                    # gsutil rsync does not support a list of files. Symlinking
+                    # all files into the same dir would also not work since
+                    # gsutil does not support following dir symlinks.
```

Comment:
Out of curiosity, why we are using `gsutil rsync` instead of `gsutil cp`? It seems that `gsutil cp` supports the following, which will significantly improve the uploading speed when multiple files are being uploaded. https://cloud.google.com/storage/docs/gsutil/commands/cp
```
cat filelist | gsutil -m cp -I gs://my-bucket
```

Reply:
Great question - `gsutil cp` will force upload a file even if it already exists on the bucket every time the task is launched. `gsutil rsync` uploads only files which either don't exist on the bucket or have been modified locally. This reduces upload time and number of operations required to sync the storage object.

Reply:
Thanks for the explanation!
The reason I mentioned this is that for the salk users, we did observe that looping through all the files to be uploaded with `gsutil rsync` or `gsutil cp` was much slower than `cat filelist | gsutil -m cp -I gs://my-bucket/subdir`(10 min vs 3 min for more than 200 files). That is to say `sky spot launch` will be blocked by the uploading of the files.
It seems `gsutil cp` supports a `-n` option that avoids overwriting the files, but if that does not have the same semantic as `gsutil rsync`, I think it should be fine to keep the current way.

Reply:
Oh that's a great point. It seems `gsutil cp -n` should do what we need. Let me try it out and update the PR if it works

Reply:
Thanks for the pointer @Michaelvll, `gsutil cp -n` indeed works similar to rsync and calling `cat files | gsutil -m cp -n -I` once is faster than calling `gsutil rsync` multiple times. Switched our implementation. PTAL